include/envoy/upstream/upstream.h

```diff
@@ -109,6 +109,11 @@ class Host : virtual public HostDescription {
    */
   virtual bool healthy() const PURE;
 
+  /**
+   * @return whether in aggregate a host is degraded and should be deprioritized during routing.
+   */
+  virtual bool degraded() const PURE;
```

Comment:
how is this conceptually different from outlier detection where a host fails repeatedly (we have success threshold/failure threshold, and other signals).
If the criteria for a degraded host is same as those gleaned from other means (failing health checks or from outlier detection), then wont it be easier to provide an API-ish way to indicate that priority routing should takeinto account failing hosts as well (albeit at lower priority)

Reply:
looking at the PR, I guess thats what you are doing? the PR description seemed to hint at an API level change (which I mistook for data plane api changes

Reply:
Yeah, maybe I should improve the title. 

The main bit in this PR is really the LB changes that makes it prioritize healthy hosts before degraded hosts. A lot of the changes end up being related to extracting the list of degraded hosts and passing those around due to the APIs changed are used all over the place.